### ISSUE TITLE
More formal definition of array index

### ISSUE THREAD

**Issue by annevk** [tc39/ecma262#900]
https://tc39.github.io/ecma262/#sec-array-exotic-objects seems to define array index somewhat informatively. Could there be a more formal definition there so it can be reused by IDL?

You probably also do not want to define it only for String properties, since as far as I can tell https://tc39.github.io/ecma262/#sec-array-exotic-objects-defineownproperty-p-desc can invoke it upon a Symbol, for which it would not be defined?

Something like an IsArrayIndex abstract operation would be great.

**Comment by anba** [comment]
_Array index_ is more formally defined in https://tc39.github.io/ecma262/#sec-object-type.

**Comment by annevk** [comment]
That still only defines it for Strings as far as I can tell.

**Comment by annevk** [comment]
For the couple of uses of integer and array index it seems using abstract operations would be much clearer overall.

**Comment by rwaldron** [comment]
From https://tc39.github.io/ecma262/#sec-object-type:

> An _array index_ is an integer index whose numeric value **_i_** is in the range +0 ≤ **_i_** < 232-1.

Did you want something more explicit, eg. with an `esid` of its own?

**Comment by annevk** [comment]
Basically something where https://tc39.github.io/ecma262/#sec-array-exotic-objects-defineownproperty-p-desc step 3 doesn't have a logic error when _P_ is a Symbol and ideally something where you can pass _P_ (even when being a Symbol) to an abstract operation IsArrayIndex() that handles all these details and returns either true or false.

**Comment by anba** [comment]
I still don't see where 9.4.2.1 contains a logic error. _Array indices_ are a subset of _integer indices_, and _integer indices_ are a subset of String-valued property keys. So every _array index_ is a String-valued property key.

**Comment by annevk** [comment]
You're right, I read it the wrong way. Still, I think an abstract operation would be much clearer.
